// vite-plugin-ssr/dist/cjs/node/plugin/plugins/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveVpsConfig = void 0;
const assertConfigVps_1 = require("./config/assertConfigVps");
const utils_1 = require("../utils");
const findConfigVpsFromStemPackages_1 = require("./config/findConfigVpsFromStemPackages");
const pickFirst_1 = require("./config/pickFirst");
const resolveExtensions_1 = require("./config/resolveExtensions");
const resolveBase_1 = require("./config/resolveBase");
function resolveVpsConfig(vpsConfig) {
    return {
        name: 'vite-plugin-ssr:resolveVpsConfig',
        enforce: 'pre',
        async configResolved(config) {
            const configVpsPromise = resolveConfigVps((vpsConfig ?? {}), (config.vitePluginSsr ?? {}), config);
            config.configVpsPromise = configVpsPromise;
        }
    };
}
exports.resolveVpsConfig = resolveVpsConfig;
async function resolveConfigVps(fromPluginOptions, fromViteConfig, config) {
    {
        const validationErr = (0, assertConfigVps_1.checkConfigVps)(fromPluginOptions);
        if (validationErr)
            (0, utils_1.assertUsage)(false, `vite.config.js > vite-plugin-ssr option ${validationErr.prop} ${validationErr.errMsg}`);
    }
    {
        const validationErr = (0, assertConfigVps_1.checkConfigVps)(fromViteConfig);
        if (validationErr)
            (0, utils_1.assertUsage)(false, `vite.config.js#vitePluginSsr.${validationErr.prop} ${validationErr.errMsg}`);
    }
    const fromStemPackages = await (0, findConfigVpsFromStemPackages_1.findConfigVpsFromStemPackages)(config.root);
    const configs = [fromPluginOptions, ...fromStemPackages, fromViteConfig];
    const { baseServer, baseAssets } = (0, resolveBase_1.resolveBase)(configs, config);
    const configVps = {
        disableAutoFullBuild: (0, pickFirst_1.pickFirst)(configs.map((c) => c.disableAutoFullBuild)) ?? false,
        extensions: (0, resolveExtensions_1.resolveExtensions)(configs, config),
        prerender: resolvePrerenderOptions(configs),
        includeAssetsImportedByServer: (0, pickFirst_1.pickFirst)(configs.map((c) => c.includeAssetsImportedByServer)) ?? false,
        baseServer,
        baseAssets
    };
    return configVps;
}
function resolvePrerenderOptions(configs) {
    if (!configs.some((c) => c.prerender)) {
        return false;
    }
    const configsPrerender = configs.map((c) => c.prerender).filter(isObject);
    return {
        partial: (0, pickFirst_1.pickFirst)(configsPrerender.map((c) => c.partial)) ?? false,
        noExtraDir: (0, pickFirst_1.pickFirst)(configsPrerender.map((c) => c.noExtraDir)) ?? false,
        parallel: (0, pickFirst_1.pickFirst)(configsPrerender.map((c) => c.parallel)) ?? true,
        disableAutoRun: (0, pickFirst_1.pickFirst)(configsPrerender.map((c) => c.disableAutoRun)) ?? false
    };
}
function isObject(p) {
    return typeof p === 'object';
}
